ruma-api-macros/src/derive_outgoing.rs

```diff
@@ -0,0 +1,184 @@
+        #vis struct #incoming_ident #ty_gen #struct_def
+
+        impl #original_impl_gen ::ruma_api::Outgoing for #original_ident #original_ty_gen {
+            type Incoming = #incoming_ident #impl_gen;
+        }
+    })
+}
+
+fn no_deserialize_in_attrs(attrs: &[Attribute]) -> bool {
+    for attr in attrs {
+        match &attr.path {
+            Path { leading_colon: None, segments }
+                if segments.len() == 1 && segments[0].ident == "incoming_no_deserialize" =>
+            {
+                return true
+            }
+            _ => {}
+        }
+    }
+
+    false
+}
+
+fn impl_outgoing_with_incoming_self(ident: Ident) -> TokenStream {
+    quote! {
+        impl ::ruma_api::Outgoing for #ident {
+            type Incoming = Self;
+        }
+    }
+}
+
+fn split_for_impl_lifetime_less(generics: &mut Generics) -> (ImplGenerics, TypeGenerics) {
+    let mut remaining_params = Punctuated::<_, Token![,]>::new();
+    for param in mem::replace(&mut generics.params, Punctuated::<_, Token![,]>::new()) {
+        if let GenericParam::Lifetime(_) = param {
+            continue;
+        } else {
+            remaining_params.push(param.clone());
+        }
+    }
+    generics.params = remaining_params;
+
+    let (impl_gen, ty_gen, _) = generics.split_for_impl();
+    (impl_gen, ty_gen)
+}
+
+fn strip_lifetimes(field: &mut Field) {
+    match &mut field.ty {
+        // T<'a> -> IncomingT
+        // The IncomingT has to be declared by the user of this derive macro.
+        Type::Path(TypePath { path, .. }) => {
+            let mut has_lifetimes = false;
+            for seg in &mut path.segments {
+                // strip generic lifetimes
+                if let PathArguments::AngleBracketed(AngleBracketedGenericArguments {
+                    args, ..
+                }) = &mut seg.arguments
+                {
+                    let mut remaining_gen = Punctuated::<_, Token![,]>::new();
+                    for arg in mem::replace(args, Punctuated::new()) {
+                        if matches!(arg, GenericArgument::Lifetime(_)) {
+                            has_lifetimes = true;
+                        } else {
+                            remaining_gen.push(arg)
+                        }
+                    }
+
+                    *args = remaining_gen;
+                }
+            }
+
+            if has_lifetimes {
+                if let Some(name) = path.segments.last_mut() {
+                    let incoming_ty_ident = format_ident!("Incoming{}", name.ident);
+                    name.ident = incoming_ty_ident;
+                }
+            }
+        }
+        Type::Reference(TypeReference { elem, .. }) => match &**elem {
+            Type::Path(ty_path) => {
+                let TypePath { path, .. } = ty_path;
+                if path.is_ident("str") {
+                    field.ty = parse_quote! { String };
+                } else if path.is_ident("ServerName") || path.is_ident("DeviceId") {
```

Comment:
This only works if `ServerName` or `DeviceId` is `use`d in the outer scope. We should probably support `ruma_identifiers::ServerName`, `ruma::identifiers::ServerName` and `ruma::ServerName` as well (plus the same for `DeviceId`, regardless of whether there is a leading `::`).